### PR TITLE
core/state/snapshot: less disk reads

### PR DESCRIPTION
**Pull request by holiman**
This PR attempts to minimize the number of disk reads. When we have a slice of snapshot values, which do not match the trie data, we currently iterates the disk to retrieve the canonical data. 
What this PR does, is commit the (incorrect) snapshot data to a trie, which will be 99% correct. When iterating the trie, we then use the snapshot-trie-database for resolving hashes. 
When doing so, we can read 99% of the leaves the from the memory db instead of resolving from disk. 

This is still work in progress, needs to be tidied up. It can also be implemented differently. The upside of this particular way to implement it is that most of the modifications are performed in the node iterator, and does not touch the trie database.

### CODE REVIEW

**Comment by holiman** [comment]
Generating without this PR:
Screenshot_2021-03-30 Dual Geth - Grafana(1)
Generating with this PR:
Screenshot_2021-03-30 Dual Geth - Grafana

^ Note that the two runs are on the same machine at different times, and the latter image shows a much shorter time segment

File: core/state/snapshot/generate.go [hunk 1]
```diff
@@ -21,6 +21,8 @@ import (
 	"encoding/binary"
 	"errors"
 	"fmt"
+	"github.com/ethereum/go-ethereum/ethdb/memorydb"
```
Comment (rjl493456442) [inline core/state/snapshot/generate.go]: Please fix it.

File: core/state/snapshot/generate.go [hunk 2]
```diff
@@ -404,16 +406,31 @@ func (dl *diskLayer) generateRange(root common.Hash, prefix []byte, kind string,
 		}
 		meter.Mark(1)
 	}
+
+	// We use the snap data to build up a cache which can be used by the
+	// main account trie as a primary lookup when resolving hashes
+	var snapTrieDb *trie.Database
+	if len(result.keys) > 0 {
+		snapNodeCache := memorydb.New()
```
Comment (rjl493456442) [inline core/state/snapshot/generate.go]: We can embed the `snapNodeCache` into the `snapTrieDb`.

```go
snapTrieDb = trie.NewDatabase(memorydb.New())
```

File: core/state/snapshot/generate.go [hunk 3]
```diff
@@ -404,16 +406,31 @@ func (dl *diskLayer) generateRange(root common.Hash, prefix []byte, kind string,
...
 		tr, err = trie.New(root, dl.triedb)
 		if err != nil {
 			return false, nil, err
 		}
 	}
+
 	var (
 		trieMore       bool
-		iter           = trie.NewIterator(tr.NodeIterator(origin))
+		nodeIt         = tr.NodeIterator(origin)
```
Comment (rjl493456442) [inline core/state/snapshot/generate.go]: I think we can pass the auxiliary database here. `iter = trie.NewIterator(tr.NodeIteratorWithDatabase(origin))`

File: core/state/snapshot/generate.go [hunk 4]
```diff
@@ -428,6 +445,7 @@ func (dl *diskLayer) generateRange(root common.Hash, prefix []byte, kind string,
 		istart   time.Time
 		internal time.Duration
 	)
+	nodeIt.AddResolver(snapTrieDb)
```
Comment (rjl493456442) [inline core/state/snapshot/generate.go]: We can pass the db directly when we construct the iterator

**Review by rjl493456442 — COMMENTED** [review]
Code wise I think it's OK, just some nitpicks

**Comment by holiman** [comment]
> Code wise I think it's OK, just some nitpicks

Yeah, well, one thing that would be kind of neat, or at least interesting to check, is to 
- not do a _second_ round of trie:ing after the proof has failed, but rather use the db from the proof-phase. Right now I wasn't sue what trie to use where, so I just created a new one for the proof-of-concept, 
- and if that works, replace the stacktrie with a regular empty trie, so we can use it if the proof is unsuccessful.

**Comment by rjl493456442** [comment]
> > Code wise I think it's OK, just some nitpicks
> 
> Yeah, well, one thing that would be kind of neat, or at least interesting to check, is to
> 
> * not do a _second_ round of trie:ing after the proof has failed, but rather use the db from the proof-phase. Right now I wasn't sue what trie to use where, so I just created a new one for the proof-of-concept,
> * and if that works, replace the stacktrie with a regular empty trie, so we can use it if the proof is unsuccessful.

Yes sure, I can do it. We need to modify the range prover a bit, let me see how to fix it.

**Comment by holiman** [comment]
Closing this in favour of https://github.com/ethereum/go-ethereum/pull/22667

**Pull request closed without merging**